[6] hansard data

beingkk opened the pull request:
Closes #6 

Added a function to fetch Hansard data from a [Zenodo repository](https://zenodo.org/record/4066772). Presently, used a very simple solution that fetches the most up to date dataset at the time of writing this PR. As the dataset is updated rather infrequently, this might be sufficient.

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

beingkk committed:
test script to fetch gtr data

beingkk committed:
function for fetching GTR data

beingkk committed:
GTR data getter

beingkk committed:
fixed test bug

beingkk committed:
check crunchbase schemas

beingkk committed:
check crunchbase schemas

beingkk committed:
crunchbase data getter

beingkk committed:
crunchbase data getter

beingkk committed:
draft readme for data inputs

beingkk committed:
draft readme for data inputs

beingkk committed:
hansard data getter

beingkk committed:
added info in the readme

jaklinger commented on innovation_sweet_spots/getters/inputs.py:
```diff
@@ -10,6 +10,7 @@
 import yaml
+import urllib.request
 
 INPUTS_PATH = f"{PROJECT_DIR}/inputs/data/"
 GTR_PATH = f"{INPUTS_PATH}gtr_projects.csv"
```
If `PROJECT_DIR` was a `pathlib.Path` object, then you could change these to:

```suggestion
INPUTS_PATH = PROJECT_DIR / "inputs/data/"
GTR_PATH = INPUTS_PATH / "gtr_projects.csv"
```

jaklinger commented on innovation_sweet_spots/getters/inputs.py:
```diff
@@ -86,7 +87,32 @@ def get_cb_data(fpath=INPUTS_PATH, cb_data_spec_path=CB_DATA_SPEC_PATH):
     return tables
 
 
+def get_hansard_data(fpath=f"{INPUTS_PATH}hansard/"):
```
```suggestion
HANSARD_PATH = INPUTS_PATH / "hansard"
...
def get_hansard_data(fpath=HANSARD_PATH):
```

jaklinger commented on innovation_sweet_spots/getters/inputs.py:
```diff
@@ -86,7 +87,32 @@ def get_cb_data(fpath=INPUTS_PATH, cb_data_spec_path=CB_DATA_SPEC_PATH):
+    fpath : str
+        Location on disk for saving the projects table
+    """
+    urls = [
+        "https://zenodo.org/record/4066772/files/government_posts.json?download=1",
+        "https://zenodo.org/record/4066772/files/opposition_posts.json?download=1",
+        "https://zenodo.org/record/4066772/files/parliamentary_posts.json?download=1",
+        "https://zenodo.org/record/4066772/files/hansard-speeches-v301.csv?download=1",
+    ]
+    logging.info(f"Collection and storing of Hansard data in progress")
+    for url in urls:
+        filename = url.split("/")[-1].split("?")[0]
+        logging.info(f"Collecting {filename}")
+        urllib.request.urlretrieve(url, f"{fpath}{filename}")
```
```suggestion
ZENODO_BASE = "https://zenodo.org/record/4066772/files{}?download=1"
ZENODO_FILES = ["government_posts.json", "opposition_posts.json", "parliamentary_posts.json", "hansard-speeches-v301.csv"]
ZENODO_URLS = map(ZENODO_BASE.format, ZENODO_FILES))

...

    filenames = ["government_posts.json", "opposition_posts.json", "parliamentary_posts.json", "hansard-speeches-v301.csv"]
    logging.info(f"Collection and storing of Hansard data in progress") 
    for filename, url in zip(ZENODO_FILES, ZENODO_URLS):
        logging.info(f"Collecting {filename}")
        urllib.request.urlretrieve(url, fpath / filename)
```

beingkk committed:
fix issue #16 make install errors

beingkk committed:
Apply suggestions from code review

Co-authored-by: Joel Klinger <joel.klinger@nesta.org.uk>

beingkk committed:
testing git-crypt

beingkk committed:
added encrypted db config for versioning

beingkk committed:
adjusted database config handling

beingkk committed:
Update README.md

beingkk committed:
Update README.md

beingkk committed:
fix to create inputs/data folder if it doesnt exist

beingkk committed:
cb data caching

beingkk committed:
add docstring to the eda notebook

beingkk committed:
Merge branch '4_access_business_data' into 6_hansard_data

beingkk committed:
Update hansard data getter

Co-authored-by: Joel Klinger <joel.klinger@nesta.org.uk>

beingkk committed:
improved Hansard getter and added support for the new version

beingkk committed:
fixing bug in tests

beingkk committed:
Merge branch '4_access_business_data' into 6_hansard_data

beingkk committed:
improving tests and fixing urls

beingkk committed:
Merge branch 'dev' into 4_access_business_data

beingkk committed:
Apply suggestions from code review

Co-authored-by: Joel Klinger <joel.klinger@nesta.org.uk>

beingkk committed:
improved use_cached logic and adjusted tests

beingkk committed:
Merge branch '4_access_business_data' into 6_hansard_data

beingkk commented:
Hi, I've updated this branch in line with the previous PRs. Note that I've designed this getter solely for downloading the data and hence I'm skipping the `use_cached` logic. This is because the table is quite large (3GB) and I was thinking that loading it in probably requires a bit more careful approach (e.g. loading via chunks; also perhaps I should create a smaller table covering more recent time period that doesn't go all the way back to 1979). Do let me know what you think.

Also, I reached out to the maintainer of the Hansard dataset and he kindly updated it to cover speeches up to April 2021. The big CSV was now uploaded as a `.zip` file on the Zenondo repo, so I added the functionality to unzip archives.

jaklinger commented on innovation_sweet_spots/getters/inputs.py:
```diff
@@ -115,10 +121,49 @@ def get_cb_data(fpath=CB_PATH, cb_data_spec_path=CB_DATA_SPEC_PATH, use_cached=T
+    for filename, url in zip(ZENODO_FILES, ZENODO_URLS):
+        logging.info(f"Collecting {filename}")
+        filepath = fpath / filename
+        urllib.request.urlretrieve(url, fpath / filename)
```
`urlretrieve` is kind of old hat (https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve) and will probably become deprecated

jaklinger replied:
It's not necessary to temporarily manage files on disk: i'd suggest streaming the data to a ~~temp file~~ [EDIT] **BytesIO** and then only extracting the file you need. Also I'd replace `urlretrieve` with a streamed request to give more stability for large files. Something like this should do it (not tested, probably needs a little tweak on the filenames or something)

```python
import requests
from io import BytesIO
import shutil
from zipfile import ZipFile


def stream_to_file(url, fileobj):
    """
    Stream a large file from a URL to a file object in a memory-efficient
    rate-efficient manner.
    """
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        if not url.endswith('zip'):  # Deflate if not a zipfile
            r.raw.decode_content = True
        shutil.copyfileobj(r.raw, fileobj)
    fileobj.seek(0)  # Reset the file pointer, ready for reading

def extract_from_zip(fileobj, fpath):
    with ZipFile(fileobj) as zf:
        zf.extract(path=fpath)

def bytesio_to_file(fileobj, fpath):
    with open(fpath, "w") as f:
        f.write(fileobj.read().decode())

def download_hansard_data(fpath=HANSARD_PATH):
    for filename, url in zip(ZENODO_FILES, ZENODO_URLS):
        with BytesIO() as fileobj:
            stream_to_file(url, fileobj)
            extract_to_disk = (extract_from_zip if filename.endswith(".zip") 
                               else (lambda obj, path: bytesio_to_file(obj, path / filename)))
            extract_to_disk(fileobj, fpath)               
```

jaklinger replied:
EDIT: I switched `TemporaryFile` to `BytesIO`: even less writing to disk

jaklinger replied:
EDIT: I saw that you also need a mode for the json file, so extended the functionality a little

jaklinger commented on inputs/README.md:
```diff
@@ -6,4 +6,8 @@ To download all input data at once you can run the following command (from the b
-To download specific input datasets, see the documentation in `inputs.py`
+To download specific input datasets, see the documentation in `inputs.py`. Presently, the datasets available via `inputs.py` are the following:
+
+- GTR research projects (approximately 400 MB)
+- Crunchbase organisations (<1GB)
+- Hansard speeches (3+GB)
```
There is a question about how much data you want to have on your local machine at any one time. It might be useful at the "read" stage (i.e. here) to already filter out columns or rows that aren't needed

jaklinger reviewed (changes requested):
Hey there - code runs fine although I got an SSL error because the zip file is so large, so needed to run twice. You can mitigate that by streaming with `requests`. I've provided some suggestions for switching to a tempfile-free setup. With the current setup you will have both the zipfile plus the extracted file on disk (at least temporarily) which will probably cause problems at some point with disk space. I've suggested moving to a `BytesIO` to hold the thing in memory before extracting the file, which should roughly halve the temporary disk space usage and doesn't rely on manual file cleanup.

One question I have is that the other PRs in this series have loaded the data ready for analysis, whereas this just does the downloading part. Is that coming up in a different PR or did you intend to put that here?

beingkk commented:
Thank you so much! Will make the adjustments you've suggested.

> One question I have is that the other PRs in this series have loaded the data ready for analysis, whereas this just does the downloading part. Is that coming up in a different PR or did you intend to put that here?

I will add this in a separate PR, once I understand better this data and know for sure what I need to pull out from there.

jaklinger commented:
@beingkk  is this good to close? Just cleaning up the ol' inbox before I leave and I can see that this is dangling
